### PR TITLE
add version property on webpack object

### PR DESCRIPTION
**Pull request by SuperManitu**
This is a very simple and convenient solution for #93

### CODE REVIEW

**Comment by andywer** [comment]
Thanks for the PR!

The solution you propose is straight-forward and I agree that `webpack.version` would be ideal (maybe the webpack authors should have added this already), but I don't think we should monkeypatch the webpack instance at all.

Imagine they add a `version` property tomorrow which has a different format...

`context.webpackVersion` might not look as nice, but such an approach may safe us a lot of future trouble 😉

I would also propose we store the parsed semver read from webpack's `package.json` instead of a hard-coded major version integer.

What do you think? Will update the PR (if I can) or create another one. Unless you disagree with my proposals. Let's settle on the "how" first :)

**Comment by SuperManitu** [comment]
sure, maybe even as object, so i can compare webpack.webpackVersion.major. Something like this:
```
webpackVersion: {
  major: 2
  minor: 1
  patch: 0
  suffix: rc.0
  string: "2.1.0-rc.0"
}
```

**Comment by andywer** [comment]
Exactly! I would say we use https://github.com/npm/node-semver, since it's a quite small dependency and the same parser `npm` uses :)

**Comment by SuperManitu** [comment]
I dont think we need an extra dependency, node-semver does a lot more than we need, I updated the core, but the tests in webpack and webpack2 are breaking without me finding the reason

**Comment by andywer** [comment]
True, `node-semver` does stuff we don't need as well. I just fear that "re-inventing the wheel" might be error-prone here, since semver allows many kinds of rare edge cases, I think.

For example the `.map(parseInt)` line: Makes sense for all versions I can remember, but I am not sure if I would go as far as saying "there will always be an integer". You know...

One more thing: `createConfig()` having two parameters before the actual setters doesn't seem nice (esp. if we add even more in future). I'd like to propose turning the first parameter into an object (`initialContext`), containing `webpack` and `webpackVersion`.

**Comment by SuperManitu** [comment]
The problem with that is that this would break all blocks that use context.webpack

**Comment by andywer** [comment]
It would not break any blocks. I am just talking about how the stuff is passed to `createConfig()`, the `context` will still look the same 😉

Like that:
```js
function createConfig (initialContext, configSetters) {
  // ...
  const context = Object.assign({}, initialContext, { fileType })
  // ...
}
```

**Comment by SuperManitu** [comment]
Ahh, ok

**Comment by andywer** [comment]
Ahh, screw my `node-semver` argumentation... After thinking about it some more I am not even sure about anymore 🙈😅

Custom version parsing code might be fine if written in a reliable way. `node-semver` is maybe not that well suited for our use case.

But I have some rather small comments on the current implementation. Will comment the code.

File: packages/core/src/index.js [hunk 1]
```diff
@@ -37,6 +39,27 @@ function createConfig (webpack, configSetters) {
...
 
 /**
+ * Parses a semver string into an object
+ * @param {string} string
+ * @return {object}
+ */
+function parseVersionString (string) {
+  const semver = string.split('-')
+  const versions = semver[0]
+    .replace(/[^0-9.]/g, '')
```
Comment (andywer) [inline packages/core/src/index.js]: Why would you `.replace(/[^0-9.]/g, '')`? That means cutting off the major version, doesn't it? 🤔

File: packages/core/src/index.js [hunk 2]
```diff
@@ -37,6 +39,27 @@ function createConfig (webpack, configSetters) {
 }
 
 /**
+ * Parses a semver string into an object
+ * @param {string} string
+ * @return {object}
+ */
+function parseVersionString (string) {
+  const semver = string.split('-')
```
Comment (andywer) [inline packages/core/src/index.js]: `const [ release, prerelease ] = string.split('-')` might be better readable

File: packages/core/src/index.js [hunk 3]
```diff
@@ -37,6 +39,27 @@ function createConfig (webpack, configSetters) {
 }
 
 /**
```
Comment (andywer) [inline packages/core/src/index.js]: I think we should move the version-parsing function into a separate file. It's not big, but it's also not related to the rest of the file's code, it's something completely independent.

**Comment by andywer** [comment]
Would you mind if I start a new feature branch based on what we discussed today?

Alternatively we can also change the code here and we squash everything before merging, so there are not so many back-and-forth commits.

**Comment by SuperManitu** [comment]
Yeah, it's easier to start a new one

**Pull request closed without merging**